src/worker/mod.rs

```diff
@@ -369,45 +394,83 @@ impl<
                 .await?;
         }
 
-        let exit = self.listen_for_heartbeats(&statuses).await;
-
-        // there are a couple of cases here:
-        //
-        //  - we got TERMINATE, so we should just return, even if a worker is still running
-        //  - we got TERMINATE and all workers have exited
-        //  - we got an error from heartbeat()
-        //
-        self.terminated = exit.is_ok();
-
-        if let Ok(true) = exit {
-            let running = self.force_fail_all_workers().await;
-            if running != 0 {
-                return Ok(running);
-            }
-        }
+        // for "forever" operations (currently only `Worker::run_to_completion`) we support SIGTERM handling,
+        // whereas for long-running operations (see `Worker::run`), we are polling the cancellation future which
+        // serves as a signal to start graceful shutdowna and return control to the calling site
+        let cancel_signal = if self.forever {
+            None
+        } else {
+            self.shutdown_signal.take()
+        };
 
-        // we want to expose worker errors, or otherwise the heartbeat error
-        let mut results = Vec::with_capacity(workers_count);
-        while let Some(res) = join_set.join_next().await {
-            results.push(res.expect("joined ok"));
-        }
+        let report = tokio::select! {
+            // A signal SIGTERM from the OS received.
+            _ = tokio::signal::ctrl_c(), if self.forever => {
+                tokio::select! {
+                    _ = tokio::signal::ctrl_c() => {
+                        process::exit(0);
+                    },
+                    _ = tokio_sleep(TokioDuration::from_millis(self.shutdown_timeout)) => {
+                        process::exit(0);
+                    },
+                    _nrunning = self.force_fail_all_workers("SIGTERM received") => {
+                        process::exit(0);
+                    }
+                }
+            },
+            // A signal from the user space received.
+            _ = async { let signal = cancel_signal.unwrap(); signal.await }, if cancel_signal.is_some() => {
+                let nrunning = tokio::select! {
+                    _ = tokio_sleep(TokioDuration::from_millis(self.shutdown_timeout)) => {
+                        0
+                    },
+                    nrunning = self.force_fail_all_workers("termination signal received over channel") => {
+                        nrunning
+                    }
+                };
+                self.terminated = true;
+                Ok(nrunning)
+            },
+           // A signal from the Faktory server received or an error occurred.
+           exit = self.listen_for_heartbeats(&statuses) => {
```

Comment:
Since this isn't cancellation-safe, we should document why we still use it in a `select!` and expect it to be okay. I think it's because if we do cancel `listen_for_heartbeats` we are exiting anyway, and the connection will never be used again?

Reply:
Exactly! Even though it  is not cancellation safe, we are ok, because we are not selecting in a loop and we are eventually _either_ marking the worker as terminated (this is being  verified as soon as we enter Worker::run, _or_ re-creating a connection (instantiating a new stream, running a handshake), i.e. we never end up re-using the "broken" connection.

Added docs.